### PR TITLE
refactor(examples) Update secure aggregation example

### PR DESCRIPTION
**Pull request by danielnugraha**
Based on #3858

### CODE REVIEW

File: examples/flower-secure-aggregation/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+---
+title: Simple Secure Aggregation with Flower Example
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:2]:
```suggestion
```

Comment (jafermarq) [reply examples/flower-secure-aggregation/README.md:2]: in the end we decided to remove the `title` from this metadata.

File: examples/flower-secure-aggregation/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+flower-secure-aggregation
+      |
+      ├── secaggexample
+      |        ├── __init__.py
+      |        ├── client_app.py    # defines your ClientApp
+      |        ├── server_app.py    # defines your ServerApp
+      |        ├── task.py
+      |        └── workflow_with_log.py
+      ├── pyproject.toml            # builds your FAB, includes dependencies and configs
+      └── README.md
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:35]:
```suggestion
flower-secure-aggregation
|
├── secaggexample
|   ├── __init__.py
|   ├── client_app.py    # Defines your ClientApp
|   ├── server_app.py    # Defines your ServerApp
|   ├── task.py
|   └── workflow_with_log.py
├── pyproject.toml       # Project metadata like dependencies and configs
└── README.md
```

File: examples/flower-secure-aggregation/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+tags: [advanced, secure_aggregation, privacy]
+dataset: [CIFAR-10]
+framework: [torch, torchvision]
+---
+
+# Secure aggregation with Flower (the SecAgg+ protocol)
+
+The following steps describe how to use Secure Aggregation in flower, with `ClientApp` using `secaggplus_mod` and `ServerApp` using `SecAggPlusWorkflowWithLogs`, which is a subclass of `SecAggPlusWorkflow` that includes more detailed logging specifically designed for this example.
+
+## Project Setup
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:12]:
```suggestion
## Set up the project

### Clone the project

```

File: examples/flower-secure-aggregation/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+## Run the Example
+
+You can run your `ClientApp` and `ServerApp` in both _simulation_ and
+_deployment_ mode without making changes to the code. If you are starting
+with Flower, we recommend you using the _simulation_ model as it requires
+fewer components to be launched manually. By default, `flwr run` will make
+use of the Simluation Engine. Refer to alternative ways of running your
+Flower application including Deployment, with TLS certificates, or with
+Docker later in this readme.
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:52]:
```suggestion
## Run the project

You can run your Flower project in both _simulation_ and _deployment_ mode without making changes to the code. If you are starting with Flower, we recommend you using the _simulation_ mode as it requires fewer components to be launched manually. By default, `flwr run` will make use of the Simulation Engine.
```

File: examples/flower-secure-aggregation/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+### Run with the Simulation Engine
+
+Run:
+
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:57]:
```suggestion
```

File: examples/flower-secure-aggregation/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+```bash
+flwr run
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:59]:
```suggestion
flwr run .
```

File: examples/flower-secure-aggregation/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+You can also override some of the settings for your `ClientApp` and `ServerApp` defined in `pyproject.toml`. For example
+
+```bash
+flwr run --run-config 'num_server_rounds=5'
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:65]:
```suggestion
flwr run . --run-config num-server-rounds=5
```

File: examples/flower-secure-aggregation/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    "flwr-datasets[vision]>=0.3.0",
+    "torch==2.2.1",
+    "torchvision==0.17.1",
+]
+
+[tool.hatch.build.targets.wheel]
+packages = ["."]
+
+[tool.flwr.app]
+publisher = "flowerteam"
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/pyproject.toml:21]:
```suggestion
publisher = "flwrlabs"
```

File: examples/flower-secure-aggregation/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,43 @@
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "secaggexample"
+version = "1.0.0"
+description = "Secure Aggregation in Flower"
+license = { text = "Apache License (2.0)" }
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/pyproject.toml:9]:
```suggestion
license = "Apache-2.0"
```

File: examples/flower-secure-aggregation/secaggexample/client_app.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
+import time
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/client_app.py:1]:
```suggestion
"""secaggexample: A Flower with SecAgg+ app."""

import time
```

File: examples/flower-secure-aggregation/secaggexample/server_app.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
+from typing import List, Tuple
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/server_app.py:1]:
```suggestion
"""secaggexample: A Flower with SecAgg+ app."""

from typing import List, Tuple
```

File: examples/flower-secure-aggregation/secaggexample/task.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
+"""From pytorchexample: A Flower / PyTorch app."""
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/task.py:1]:
```suggestion
"""secaggexample: A Flower with SecAgg+ app."""
```

File: examples/flower-secure-aggregation/secaggexample/__init__.py [hunk 13]
```diff
@@ -0,0 +1 @@
+"""secaggexample."""
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/__init__.py:1]:
```suggestion
"""secaggexample: A Flower with SecAgg+ app."""
```

File: examples/flower-secure-aggregation/secaggexample/workflow_with_log.py [hunk 14]
```diff
@@ -1,14 +1,19 @@
+from logging import DEBUG, INFO
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/workflow_with_log.py:1]:
```suggestion
"""secaggexample: A Flower with SecAgg+ app."""

from logging import DEBUG, INFO
```

File: examples/flower-secure-aggregation/secaggexample/server_app.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    # Multiply accuracy of each client by number of examples used
+    accuracies = [num_examples * m["accuracy"] for num_examples, m in metrics]
+    examples = [num_examples for num_examples, _ in metrics]
+
+    # Aggregate and return custom metric (weighted average)
+    return {"accuracy": sum(accuracies) / sum(examples)}
+
+
+ndarrays = get_weights(Net())
+parameters = ndarrays_to_parameters(ndarrays)
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/server_app.py:23]: Let's better move it inside `@app.main()` 🙏

File: examples/flower-secure-aggregation/secaggexample/server_app.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        fraction_fit=1.0,
+        # Disable evaluation in demo
+        fraction_evaluate=0.0 if IS_DEMO else context.run_config["fraction-evaluate"],
+        min_available_clients=5,
+        evaluate_metrics_aggregation_fn=weighted_average,
+        initial_parameters=parameters,
+    )
+
+    # Construct the LegacyContext
+    num_rounds = int(context.run_config["num-server-rounds"])
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/server_app.py:44]:
```suggestion
    num_rounds = context.run_config["num-server-rounds"]
```

File: examples/flower-secure-aggregation/README.md [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+      |        ├── client_app.py    # defines your ClientApp
+      |        ├── server_app.py    # defines your ServerApp
+      |        ├── task.py
+      |        └── workflow_with_log.py
+      ├── pyproject.toml            # builds your FAB, includes dependencies and configs
+      └── README.md
+```
+
+## Install dependencies
+
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:39]:
```suggestion
### Install dependencies and project

Install the dependencies defined in `pyproject.toml` as well as the `secaggexample` package.

```

File: examples/flower-secure-aggregation/README.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+```bash
+pip install .
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/README.md:41]:
```suggestion
pip install -e .
```

**Review by jafermarq — COMMENTED** [review]
@panh99 @danielnugraha most of my comments below are just to align the style of the `README.md` to those of the examples already migrated (e.g. `quickstart-mlx/pytorch/tensorflow/fastai`)

File: examples/flower-secure-aggregation/secaggexample/server_app.py (hunk 15 above)
Comment (panh99) [reply examples/flower-secure-aggregation/secaggexample/server_app.py]: Sure!

File: examples/flower-secure-aggregation/secaggexample/task.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        "val_accuracy": val_acc,
+    }
+    return results
+
+
+def test(net, testloader, device):
+    """Validate the model on the test set."""
+    if is_demo:
+        return 0.0, 0.0
+    criterion = torch.nn.CrossEntropyLoss()
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/task.py:123]:
```suggestion
    net.to(device)  # move model to GPU if available
    criterion = torch.nn.CrossEntropyLoss()
```

Comment (jafermarq) [reply examples/flower-secure-aggregation/secaggexample/task.py:123]: in `train()` we send it to device. I think it's better to keep the same behaviour also in `test()`.

File: examples/flower-secure-aggregation/secaggexample/client_app.py [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+        """Evaluate the model on the data this client has."""
+        set_weights(self.net, parameters)
+        loss, accuracy = test(self.net, self.valloader, self.device)
+        return loss, len(self.valloader.dataset), {"accuracy": accuracy}
+
+
+def client_fn(context: Context):
+    """Construct a Client that will be run in a ClientApp."""
+    # Set `task.is_demo`
+    task.is_demo = context.run_config["is-demo"]
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/client_app.py:55]: Would it be better to pass this `is_demo` to `load_data()` and then to the client so it skips calling `train()` and `test()`?

File: examples/flower-secure-aggregation/secaggexample/client_app.py [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+            self.device,
+        )
+        ret_vec = get_weights(self.net)
+
+        # Force a significant delay for testing purposes
+        if "drop" in config and config["drop"]:
+            print(f"Client dropped for testing purposes.")
+            time.sleep(15)
+        else:
+            print(f"Client uploading parameters: {ret_vec[0].flatten()[:3]}...")
```
Comment (jafermarq) [inline examples/flower-secure-aggregation/secaggexample/client_app.py:42]: Could we only do this if `is_demo` is `True`?

**Review by jafermarq — COMMENTED** [review]
Looks pretty good ! I just wonder if we could "drop" the client by raising an exception instead of adding a long sleep. wdty?

**Review by jafermarq — COMMENTED** [review]
Looks pretty good ! I just wonder if we could "drop" the client by raising an exception instead of adding a long sleep. wdty?

File: examples/flower-secure-aggregation/secaggexample/task.py (hunk 19 above)
Comment (panh99) [reply examples/flower-secure-aggregation/secaggexample/task.py:123]: Ok, sure. I copied `task.py` from the quickstart-pytorch example, so maybe we also want to update the `test` function in that example.

Comment (jafermarq) [reply examples/flower-secure-aggregation/secaggexample/task.py:123]: yes. will do. good point

File: examples/flower-secure-aggregation/secaggexample/client_app.py (hunk 20 above)
Comment (panh99) [reply examples/flower-secure-aggregation/secaggexample/client_app.py:55]: The client has skipped `train` and `test` under the hood. In `task.py`, when `is_demo` is `True`, the `train` and `test` functions will directly return a fake result.

**Review by jafermarq — APPROVED** [review]
@danielnugraha @panh99 Amazing update to this example!
